File: device/src/network/dns.rs
//! DNS hijack server — responds to all queries with the gateway IP.
//!
//! Pure functions compile on all targets.
//! The async `dns_hijack` task is gated behind `#[cfg(feature = "portal")]`.

use heapless::Vec as HVec;

/// Gateway IP that ALL DNS queries resolve to in AP mode.
pub const GATEWAY: [u8; 4] = [192, 168, 4, 1];

/// Find the end of a DNS QNAME starting at `offset`.
/// QNAME is a sequence of length-prefixed labels ending with a 0 byte.
/// Returns the offset of the byte AFTER the terminating 0.
pub fn find_qname_end(data: &[u8], mut offset: usize) -> Option<usize> {
    loop {
        if offset >= data.len() {
            return None;
        }
        let label_len = data[offset] as usize;
        if label_len == 0 {
            return Some(offset + 1);
        }
        // Pointer (compression) — shouldn't appear in queries, but handle it
        if label_len >= 0xC0 {
            return Some(offset + 2);
        }
        offset += 1 + label_len;
    }
}

/// Build a DNS A-record response for any query.
/// Copies the transaction ID and question from the query,
/// appends a single A record pointing to GATEWAY.
pub fn build_dns_response(query: &[u8]) -> Option<HVec<u8, 512>> {
    // Minimum DNS query: 12 byte header + at least 1-byte name + 4 bytes QTYPE/QCLASS
    if query.len() < 17 {
        return None;
    }

    let mut resp = HVec::<u8, 512>::new();

    // Transaction ID (bytes 0-1)
    resp.extend_from_slice(&query[0..2]).ok()?;

    // Flags: standard response, recursion available, no error
    resp.extend_from_slice(&[0x81, 0x80]).ok()?;

    // Question count (bytes 4-5, copied from query)
    resp.extend_from_slice(&query[4..6]).ok()?;

    // Answer count: 1
    resp.extend_from_slice(&[0x00, 0x01]).ok()?;

    // Authority RR count: 0
    resp.extend_from_slice(&[0x00, 0x00]).ok()?;

    // Additional RR count: 0
    resp.extend_from_slice(&[0x00, 0x00]).ok()?;

    // Question section — copy verbatim from query
    let qname_end = find_qname_end(query, 12)?;
    let question_end = qname_end + 4; // +2 QTYPE + 2 QCLASS
    if question_end > query.len() {
        return None;
    }
    resp.extend_from_slice(&query[12..question_end]).ok()?;

    // Answer section: A record pointing to gateway
    resp.extend_from_slice(&[
        0xC0, 0x0C, // Name: pointer to question name at offset 12
        0x00, 0x01, // Type: A
        0x00, 0x01, // Class: IN
        0x00, 0x00, 0x00, 0x3C, // TTL: 60 seconds
        0x00, 0x04, // Data length: 4 (IPv4)
    ])
    .ok()?;
    resp.extend_from_slice(&GATEWAY).ok()?;

    Some(resp)
}

#[cfg(feature = "portal")]
#[embassy_executor::task]
pub async fn dns_hijack(stack: embassy_net::Stack<'static>) {
    use embassy_net::udp::{PacketMetadata, UdpSocket};
    use esp_println::println;

    let mut rx_meta = [PacketMetadata::EMPTY; 4];
    let mut tx_meta = [PacketMetadata::EMPTY; 4];
    let mut rx_buf = [0u8; 512];
    let mut tx_buf = [0u8; 512];

    let mut socket = UdpSocket::new(stack, &mut rx_meta, &mut rx_buf, &mut tx_meta, &mut tx_buf);
    socket.bind(53).unwrap();
    println!("DNS hijack server started on port 53");

    let mut query_buf = [0u8; 512];
    loop {
        let (n, remote) = match socket.recv_from(&mut query_buf).await {
            Ok(result) => result,
            Err(_) => continue,
        };

        if let Some(response) = build_dns_response(&query_buf[..n]) {
            let _ = socket.send_to(&response, remote).await;
        }
    }
}

#[cfg(test)]
#[path = "dns_tests.rs"]
mod tests;

File: esp-idf/main/network/dns.c
#include "dns.h"
#include <string.h>

int dns_find_qname_end(const uint8_t *data, size_t data_len, size_t offset) {
    while (1) {
        if (offset >= data_len) {
            return -1;
        }
        uint8_t label_len = data[offset];
        if (label_len == 0) {
            return (int)(offset + 1);
        }
        // Pointer (compression)
        if (label_len >= 0xC0) {
            return (int)(offset + 2);
        }
        offset += 1 + label_len;
    }
}

int dns_build_response(const uint8_t *query, size_t query_len,
                       uint8_t *resp_buf, size_t resp_buf_size,
                       const uint8_t gateway_ip[4]) {
    // Minimum DNS query: 12-byte header + 1-byte name + 4-byte QTYPE/QCLASS
    if (query_len < 17) {
        return -1;
    }

    int qname_end = dns_find_qname_end(query, query_len, 12);
    if (qname_end < 0) {
        return -1;
    }

    size_t question_end = (size_t)qname_end + 4; // +2 QTYPE + 2 QCLASS
    if (question_end > query_len) {
        return -1;
    }

    size_t question_len = question_end - 12;

    // Response = 12 header + question + 16 answer (2 name ptr + 2 type + 2 class + 4 TTL + 2 rdlen + 4 IP)
    size_t resp_len = 12 + question_len + 16;
    if (resp_len > resp_buf_size) {
        return -1;
    }

    size_t pos = 0;

    // Transaction ID (bytes 0-1)
    resp_buf[pos++] = query[0];
    resp_buf[pos++] = query[1];

    // Flags: standard response, recursion available
    resp_buf[pos++] = 0x81;
    resp_buf[pos++] = 0x80;

    // Question count (from query)
    resp_buf[pos++] = query[4];
    resp_buf[pos++] = query[5];

    // Answer count: 1
    resp_buf[pos++] = 0x00;
    resp_buf[pos++] = 0x01;

    // Authority RR count: 0
    resp_buf[pos++] = 0x00;
    resp_buf[pos++] = 0x00;

    // Additional RR count: 0
    resp_buf[pos++] = 0x00;
    resp_buf[pos++] = 0x00;

    // Question section (copied from query)
    memcpy(resp_buf + pos, query + 12, question_len);
    pos += question_len;

    // Answer section: A record
    resp_buf[pos++] = 0xC0; // Name pointer to offset 12
    resp_buf[pos++] = 0x0C;
    resp_buf[pos++] = 0x00; // Type: A
    resp_buf[pos++] = 0x01;
    resp_buf[pos++] = 0x00; // Class: IN
    resp_buf[pos++] = 0x01;
    resp_buf[pos++] = 0x00; // TTL: 60 seconds
    resp_buf[pos++] = 0x00;
    resp_buf[pos++] = 0x00;
    resp_buf[pos++] = 0x3C;
    resp_buf[pos++] = 0x00; // Data length: 4
    resp_buf[pos++] = 0x04;

    // Gateway IP
    resp_buf[pos++] = gateway_ip[0];
    resp_buf[pos++] = gateway_ip[1];
    resp_buf[pos++] = gateway_ip[2];
    resp_buf[pos++] = gateway_ip[3];

    return (int)pos;
}
